### PR TITLE
Reactive isAlive

### PR DESCRIPTION
**Pull request by xaviergonz**
This PR makes isAlive reactive, useful if you want to react when a node becomes dead (for example by setting it to undefined or whatever)

### CODE REVIEW

**Comment by xaviergonz** [comment]
Maybe I'm wrong, but I think if isAlive is used outside an "observed" chunk it won't actually consume more memory, like a computed does?

Also, it saves you from making the mistake of using it inside an observer render/reaction/autorun and leave you wondering why the component is not re-rendering / the reaction is not running when the node dies

File: packages/mobx-state-tree/src/core/node/BaseNode.ts [hunk 1]
```diff
@@ -15,9 +15,23 @@ import { createAtom } from "mobx"
...
+    get state() {
+        return this._state
+    }
+    set state(val: NodeLifeCycle) {
+        const wasAlive = this.isAlive
+        this._state = val
+        const isAlive = this.isAlive
+
+        if (wasAlive !== isAlive) {
+            this.aliveAtom.reportChanged()
```
Comment (k-g-a) [inline packages/mobx-state-tree/src/core/node/BaseNode.ts]: I'm not sure if it's safe to call ```reportChanged()``` if you did not ```reportObserved()``` before?
I mean if you had not used ```observableIsAlive``` and node became ```DEAD``` so we hit this code: wouldnt it accidantely throw? It's just a suggestion, as I am not aquainted with mobx internals deep enough.

Comment (xaviergonz) [reply packages/mobx-state-tree/src/core/node/BaseNode.ts]: I think it is ok (or else a gazillion unit tests would throw) but I'll change it so the atom is only created when the observable is alive is used and that way we kill two birds with one stone

**Comment by xaviergonz** [comment]
Added 3 small optimizations:
- made is alive atom creation lazy
- made path atom creation lazy
- made escape path generation lazy

File: packages/mobx-state-tree/src/core/node/BaseNode.ts (hunk 1 above)
Comment (mweststrate) [reply packages/mobx-state-tree/src/core/node/BaseNode.ts]: `reportChanged()` and `reportObserved()` are allowed to be called without any further assumptions on the other to be called

File: packages/mobx-state-tree/src/core/node/BaseNode.ts [hunk 2]
```diff
@@ -6,40 +6,77 @@ import {
...
         this.environment = environment
         this.type = type
-        this.parent = parent
-        this.subpath = subpath
-        this.escapedSubpath = escapeJsonPath(this.subpath)
+        this.partialSetParent(parent, subpath)
+    }
+
+    private pathAtom?: IAtom
+    protected partialSetParent(parent: ObjectNode | null, subpath: string) {
```
Comment (mweststrate) [inline packages/mobx-state-tree/src/core/node/BaseNode.ts]: note the most descriptive name. What is "partial" about it? Probably deserves a clearer name or a comment :)

Comment (xaviergonz) [reply packages/mobx-state-tree/src/core/node/BaseNode.ts]: That there's an abstract setParent, what's a good prefix for internal functions? An underscore? Internal? I'm open to suggestions

Comment (mweststrate) [reply packages/mobx-state-tree/src/core/node/BaseNode.ts]: I typically use base as well for such cases indeed

Comment (xaviergonz) [reply packages/mobx-state-tree/src/core/node/BaseNode.ts]: sounds good, renamed partialXXX to baseXXX

**Pull request merged**